### PR TITLE
Add LIGO LDAP credentials (OPS-122)

### PR DESCRIPTION
**Pull request by brianhlin**
I went with this to avoid touching too much code. Not sure if you guys prefer reading the passfile in `models.py`

### CODE REVIEW

File: src/stashcache.py [hunk 1]
```diff
@@ -178,19 +179,24 @@ def _cache_is_allowed(resource, vo_name, stashcache_data, public, suppress_error
     return ret
 
 
-def generate_cache_authfile(vo_data: VOsData, resource_groups: List[ResourceGroup], fqdn=None, legacy=True, suppress_errors=True) -> str:
+def generate_cache_authfile(global_data: GlobalData,
```
Comment (edquist) [inline src/stashcache.py:182]: Do you need to give similar treatment to `generate_public_cache_authfile` also, to handle the `if public_only` block here?

https://github.com/opensciencegrid/topology/blob/e2d34ad34579f974c5467dc93617d8bd299fe30a/src/app.py#L253-L256

Comment (brianhlin) [reply src/stashcache.py:182]: Yup! I'll get right on it

**Comment by brianhlin** [comment]
Force pushed to fix a few issues and rebase on master

**Comment by brianhlin** [comment]
Ignore the base branch swaps. I just wanted to test out the new GitHub functionality to change base branches and @edquist had questions about changes in wording :)

File: src/tests/verify_authfile.sh [hunk 2]
```diff
@@ -2,4 +2,4 @@
 
 cd "$(dirname "$0")/../../"
 export PYTHONPATH=src/:$PYTHONPATH
-exec ./bin/osg-authfile > /dev/null
+exec ./bin/osg-authfile > /dev/null 2>&1
```
Comment (edquist) [inline src/tests/verify_authfile.sh:5]: So, you want to hide a traceback when there are errors?

Comment (brianhlin) [reply src/tests/verify_authfile.sh:5]: Yeah, AIUI, these tracebacks would be logged but since we're running the functions manually, they get dumped to `stderr`. The functions themselves handle these particular exceptions and so they exit zero, we just want to throw away their out (I forget why, @matyasselmeci ?)  and err because GitHub Actions apparently look for these tracebacks and "annotate" the test run with an error but otherwise lets the test pass because the script exist zero. See https://github.com/brianhlin/topology/actions/runs/1144133515. Exceptions that we aren't catching should still cause the script to exit non-zero.

**tl;dr** yeah, we want to throw these tracebacks away since we catch them otherwise they just lead to confusing GHA results

Comment (edquist) [reply src/tests/verify_authfile.sh:5]: Well i admit i am confused.

But let's see.  IIUC:

- there are uncaught exceptions
- these produce tracebacks on stderr
- GHA uses stderr output to annotate results
- these GHA runs still pass because (for some reason the script) still exits `0`
- we throw out stderr entirely to avoid the confusing GHA annotations with such "passing" jobs
- uncaught exceptions should still cause the script to exit non-zero

So, for this guy,

> ```Error:     raise LDAPPasswordIsMandatoryError('password is mandatory in simple bind')```

is this an exception we catch & handle, but print the traceback anyway, or is it an uncaught exception that _should_ cause the script to exit non-zero, but doesn't?

**Comment by brianhlin** [comment]
> is this an exception we catch & handle, but print the traceback anyway, or is it an uncaught exception that should cause the script to exit non-zero, but doesn't?

The former since `LDAPPasswordIsMandatoryError` is a child of `LDAPException`. It's being dumped to stderr because we use `log.exception()` without a logfile destination in the case that these functions are being called directly.

> these GHA runs still pass because (for some reason the script) still exits 0

We ignore these errors because we still can produce a partial stash cache/origin authfile without the info from LDAP. So if we encounter this, we just want to log and move on.

**Comment by edquist** [comment]
Ok, thank you for the explanation.

I guess the only downside then is you also lose the stderr from any uncaught exceptions.  They will still fail (exit non-zero) as you said, but they'd be harder to diagnose without the traceback.

**Comment by matyasselmeci** [comment]
> We ignore these errors because we still can produce a partial stash cache/origin authfile without the info from LDAP.

We may want to revisit this idea at some point -- I don't like the thought of service authfiles going back and forth between "LIGO" and "no LIGO" versions in case their LDAP server is flakey.

How about having separate endpoints for the LIGO and the not-LIGO parts of the config and having the authfile-update timer combine the two? We already have code for combining authfiles here: https://github.com/opensciencegrid/xcache/pull/117/files

**Comment by brianhlin** [comment]
I think that's a good idea but a separate endpoint + combining authfiles would still have you oscillating between LIGO/no LIGO in the case of flaky LDAP. The only way around that I can think of is keeping a LIGO authfile cache around and using that if LDAP queries fail.

Can we handle that in a different ticket/PR, though?

**Comment by matyasselmeci** [comment]
Oh yeah totally a different PR.

It wouldn't go back and forth -- like what it does with the Authfile, authfile-update will keep its previous copy of Authfile-ligo if downloading from the LIGO endpoint fails. (So yeah it's sorta like a cache, except client-side.)

File: src/tests/verify_authfile.sh (hunk 2 above)
Comment (brianhlin) [reply src/tests/verify_authfile.sh:5]: Oops, I broke off the comments into the main PR thread. See that for the explanation

**Review by edquist — APPROVED** [review]
Godspeed

**Comment by brianhlin** [comment]
Force pushed to rebase on master
